### ISSUE TITLE
1-shot spoof

### ISSUE THREAD

**Issue by JackLin-Authme** [Pei-KaiHuang/CVPR24_OC_SCMNet#1]
First, thank you for your great work.
I have a question about 1-shot Spoof.
Where is the code of it? I would like to know how you train the 1-shot.

**Comment by Pei-KaiHuang** [comment]
Hi, Lin. 

Thank you for your interest in our work. Because the students are busy preparing for graduation, we will upload the 1-shot part of the code as soon as possible afterward. 

In addition, we suggest that you can modify the existing code slightly to make it into a one-shot version.  We describe the detail in the paper: "To conduct this experiment, we randomly select one spoof image from the training dataset and additionally include a loss term to maximize the similarity between the generated latent spoof features and the extracted latent spoof feature of this spoof image."

In particular, we first set the SCM of one specific spoof image to cover the whole spatial region (i.e., SCM set to 255). Next, we use cosine similarity loss to guide the generated spoof features to be similar to the actual ones.  The training procedures can refer to the one-class code.

Sincerely, I hope I can be of help to you.

**Comment by JackLin-Authme** [comment]
Thank you for your assistance. I will follow the instruction.

**Comment by umitkacar** [comment]
Hi @Pei-KaiHuang 

I really like and appreciate your work. 

After carefully reviewing your instructions and the available codebase, I attempted to implement the one-shot spoofing procedure myself. Below is a summary of the modifications I made, as well as the code snippets for your review:

1.  I integrated an additional training phase to select a single spoof image and generate its corresponding Spoof Cue Map (SCM).
2. The SCM was set to 255 for all spatial regions, as per your guidance.
3. I introduced a cosine similarity loss term to maximize the similarity between the generated latent spoof features and the selected spoof image's latent spoof features.
4. This additional loss term was combined with the existing training loss functions during the learning of the generator (G).

```
# Generate a single spoof SCM with all spatial regions set to 255
spoof_scm = torch.ones_like(live_map) * 255  # Assuming live_map provides the shape
# Extract the latent feature for the single selected spoof image
spoof_image = torch.randn(1, 3, 256, 256).to(device)  # Placeholder for the selected spoof image
spoof_feature = model.F(NormalizeData_torch(spoof_image))  # NormalizeData_torch normalizes the input

# Generate latent spoof features using the generator
generated_spoof_feature = model.G(torch.cat([spoof_feature, spoof_scm], dim=1))

# Compute cosine similarity loss
cos_loss = -torch.mean(COS(generated_spoof_feature, spoof_feature))

# Combine with existing loss functions
final_loss = loss_g + cos_loss * 0.1  # Adjust weight as appropriate

# Backpropagation
optimizer.zero_grad()
final_loss.backward()
optimizer.step()
```
Could you kindly confirm if this approach aligns with your intended implementation of one-shot spoofing as described in the paper? Additionally, if there are specific adjustments or refinements needed, I would greatly appreciate your feedback.

I understand that the one-shot code is currently not included in the repository due to pending commitments. However, if you could release even a partial implementation or additional guidance, it would be incredibly helpful for ensuring my approach adheres to the intended methodology.

Thank you for your time and contributions to this important work. I look forward to your insights and guidance.

**Comment by umitkacar** [comment]
Hi @JackLin-Authme 

Were you able to complete the one shot spoof?

**Comment by JackLin-Authme** [comment]
Hi @umitkacar 

No, I have not tried the one-shot setting.
